Update Product Details block to use WordPress core Accordion in WP 6.9+

AbdullahAymanMSRE opened the pull request:
Fixes #62120 

- Add version detection for WordPress 6.9+
- Update block template to use core/accordion blocks when available
- Maintain backward compatibility with WooCommerce accordion for older WP versions
- Update PHP rendering to conditionally use appropriate block names"

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This PR updates the Product Details block to use the WordPress core core/accordion block when running WordPress 6.9 or newer, while fully maintaining backward compatibility with the existing WooCommerce accordion block for older WordPress versions.

This resolves issue [#62120](https://github.com/woocommerce/woocommerce/issues/62120).

Closes #62120.

(For Bug Fixes) Bug introduced in PR # .

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Install WordPress 6.9+.
2. Create/edit a product and insert a Product Details block.
3. Confirm the block uses the core `core/accordion` block.
4. Switch to WordPress 6.8 or earlier.
5. Insert the Product Details block again.
6. Confirm it falls back to the WooCommerce accordion implementation.
7. Verify that saving, editing, and rendering work on both versions without errors.

<!-- End testing instructions -->

### Testing that has already taken place:

- Tested on WordPress 6.9 nightly and WordPress 6.8.
- Verified the accordion renders correctly in both versions.
- Confirmed there are no block validation errors.
- Ran the test: tests/php/src/Blocks/BlockTypes/ProductDetails/ProductDetails.php

### Milestone

<!-- DO NOT remove or modify this section (other than to check the box). -->

<!-- milestone-target-selection -->
- [x] Automatically assign milestone for the **[next WooCommerce version](../blob/trunk/plugins/woocommerce/woocommerce.php#L6)**
<!-- /milestone-target-selection -->

> **Note:** Check the box above to have the milestone automatically assigned when merged.
> Alternatively (e.g. for point releases), manually assign the appropriate milestone.


### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [x] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->
Use WordPress core Accordion block in Product Details block for WP 6.9+, with fallback for older versions.

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

AbdullahAymanMSRE committed:
Update Product Details block to use WordPress core Accordion in WP 6.9+

Fixes #62120

- Add version detection for WordPress 6.9+
- Update block template to use core/accordion blocks when available
- Maintain backward compatibility with WooCommerce accordion for older WP versions
- Update PHP rendering to conditionally use appropriate block names"

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

Aljullu commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductDetails.php:
```diff
@@ -1,15 +1,19 @@
  */
-class ProductDetails extends AbstractBlock {
+class ProductDetails extends AbstractBlock
+{
```
We need to keep this formatting as it was, otherwise Lint automations are going to fail.

Aljullu commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductDetails.php:
```diff
@@ -34,11 +39,49 @@ protected function initialize() {
+	 */
+	private function get_accordion_block_names()
+	{
+		if ($this->is_wp_69_or_later()) {
```
Like CodeRabbit says above, we should use the util that already exists: `Utils::wp_version_compare( '6.9', '>=' )`, like [here](https://github.com/woocommerce/woocommerce/blob/3a57a10fb346e02aab1c6cecfb9a8064ec7d275a/plugins/woocommerce/src/Admin/Features/ProductBlockEditor/BlockRegistry.php#L206).

Aljullu commented on pnpm-lock.yaml:
I would exclude any changes on `pnpm-lock.yaml` from this PR, as they are unrelated.

Aljullu commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-details/utils.ts:
```diff
@@ -2,9 +2,42 @@
+ * Get accordion block names based on WordPress version
+ */
+const getAccordionBlockNames = () => {
+	if ( isWP69OrLater() ) {
```
We have a JS util for this too, like it's used [here](https://github.com/woocommerce/woocommerce/blob/3a57a10fb346e02aab1c6cecfb9a8064ec7d275a/packages/js/block-templates/src/utils/register-woo-block-type.ts#L89), so you shouldn't need the custom function. 🙂

Aljullu reviewed (commented):
Thanks for working on this, @AbdullahAymanMSRE! 🙌 Overall it's looking good, but I left some early comments below.

There are also some issues with the formatting that will make Lint fail. It's likely you have some settings in your code editor that automatically made all those changes. We would need them to be removed in order to be able to merge this PR. Could you take a look? 🙏 You can see where the errors are by running `pnpm --filter=@woocommerce/plugin-woocommerce lint:php`.

AbdullahAymanMSRE committed:
Address code review feedback

- Fix PHP linting issues and formatting
- Remove unintended whitespace changes
- Use already existing utils for checking wp version

AbdullahAymanMSRE committed:
Address code review feedback

- Revert pnpm-lock to the original content

AbdullahAymanMSRE committed:
Fix PHP linting issue by adding missing newline at end of file in ProductDetails class

Aljullu reviewed (commented):
Thanks for the updates, @AbdullahAymanMSRE! I did some testing, and it's working great! There are some subtle styling differences between the two blocks, but I believe they are desirable, considering it brings us closer to the upstream block and styles:

Before | After
--- | ---
<img width="1721" height="547" alt="Image" src="https://github.com/user-attachments/assets/073060f1-87e1-498d-87fc-83edcca2cc92" /> | <img width="1721" height="547" alt="Image" src="https://github.com/user-attachments/assets/6b5d6652-93dd-4109-900f-806e48370d2d" />

There are still a few changes we need to make to the PR in order to be able to merge it:

* There are some [PHP Linting warnings](https://github.com/woocommerce/woocommerce/actions/runs/20121870736/job/57771397367?pr=62339), they should be easy to fix, as they are about code styling.
* Integration tests are failing because the WP version is missing. So we will need to mock the `isWpVersion` function in the [_Product Details_ integration tests](https://github.com/woocommerce/woocommerce/blob/d664ee82879b8ef5fd6c3b714276b2f743110081/plugins/woocommerce/client/blocks/assets/js/blocks/product-details/test/block.tsx#L56). Something like this.

```JS
jest.mock( '@woocommerce/settings', () => ( {
	...jest.requireActual( '@woocommerce/settings' ),
	isWpVersion: jest.fn().mockReturnValue( false ),
} ) );
```

AbdullahAymanMSRE commented:
Thank you for the review @Aljullu , okay I will fix that. Sorry for that, it's my first contribution, so I am still getting used to how things are done here.
Thanks for your patience.

AbdullahAymanMSRE committed:
Fix tests for Product Details block

Aljullu commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductDetails.php:
```diff
@@ -191,7 +208,8 @@ private function create_accordion_item_block( $title, $content ) {
 	private function inject_parsed_accordion_blocks( $parsed_block, $accordion_blocks ) {
-		if ( 'woocommerce/accordion-group' === $parsed_block['blockName'] ) {
+		$accordion_group_name = Utils::wp_version_compare( '6.9', '>=' ) ? 'core/accordion' : 'woocommerce/accordion-group';
+		if ( $accordion_group_name === $parsed_block['blockName'] ) {
```
I would prefer if in this `if`, we check for both blocks, no matter the version of WordPress. That would cover the case of somebody having added the block in WP 6.8 and then upgrading to WP 6.9. The old accordions would still be used, even if the store is in WP 6.9.

```suggestion
		$accordion_group_name = Utils::wp_version_compare( '6.9', '>=' ) ? 'core/accordion' : 'woocommerce/accordion-group';
		if ( 'core/accordion' === $parsed_block['blockName'] || 'woocommerce/accordion-group' === $parsed_block['blockName'] ) {
```

Aljullu commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductDetails.php:
```diff
@@ -220,11 +238,13 @@ private function inject_parsed_accordion_blocks( $parsed_block, $accordion_block
 		}
 
-		if ( 'woocommerce/accordion-group' === $parsed_block['blockName'] ) {
+		if ( $accordion_group_name === $parsed_block['blockName'] ) {
```
Same here, I would update this check for: `'core/accordion' === $parsed_block['blockName'] || 'woocommerce/accordion-group' === $parsed_block['blockName']`. 🙂

Aljullu commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductDetails.php:
```diff
@@ -280,7 +300,9 @@ private function mark_accordion_item_hidden( $item, $context ) {
-		if ( 'woocommerce/accordion-group' === $parsed_block['blockName'] ) {
+		$accordion_group_name = Utils::wp_version_compare( '6.9', '>=' ) ? 'core/accordion' : 'woocommerce/accordion-group';
+
+		if ( $accordion_group_name === $parsed_block['blockName'] ) {
```
And here.

Aljullu commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductDetails.php:
```diff
@@ -357,11 +379,13 @@ private function validate_hooked_blocks( $hooked_blocks ) {
+			function ( $hooked_block_types, $relative_position, $anchor_block_type ) use ( $slug, $accordion_group_name ) {
 				if (
-					'woocommerce/accordion-group' === $anchor_block_type &&
+					$accordion_group_name === $anchor_block_type &&
```
And also in this one...

Aljullu commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductDetails.php:
```diff
@@ -375,10 +399,10 @@ function ( $hooked_block_types, $relative_position, $anchor_block_type ) use ( $
 				if (
 					is_null( $parsed_hooked_block ) ||
-					'woocommerce/accordion-group' !== $parsed_anchor_block['blockName'] ||
+					$accordion_group_name !== $parsed_anchor_block['blockName'] ||
```
... and this.

Aljullu reviewed (commented):
> Thank you for the review @Aljullu , okay I will fix that. Sorry for that, it's my first contribution, so I am still getting used to how things are done here.
> Thanks for your patience.

No worries, thank _you_ for your contribution. It's much appreciated! 🙌 

I left a few more comments that I missed in my first review. I believe in some places we need to check for the presence of both accordion blocks, independently of the WordPress version. Sorry for missing that initially!

I also noticed we will need to update one extra test. The new accordion block includes the `+` and `x` symbols in the label, so our existing tests fail. Basically, we need to update this locator [here](https://github.com/woocommerce/woocommerce/blob/be052facca419be9e9b1a14c9b1679690091395e/plugins/woocommerce/client/blocks/tests/e2e/tests/add-to-cart-with-options/add-to-cart-with-options.block_theme.spec.ts#L194) and [here](https://github.com/woocommerce/woocommerce/blob/be052facca419be9e9b1a14c9b1679690091395e/plugins/woocommerce/client/blocks/tests/e2e/tests/add-to-cart-with-options/add-to-cart-with-options.block_theme.spec.ts#L214) for them to pass again: 🤞 

```diff
- .getByLabel( 'Additional Information', { exact: true } )
+ .getByRole( 'region', {name: 'Additional Information' } )
```

Thanks in advance!

dinhtungdu commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductDetails.php:
```diff
@@ -162,22 +163,38 @@ function ( $a, $b ) {
+					</div>
+					<!-- /wp:core/accordion-panel -->
+				</details>
+				<!-- /wp:core/accordion-item -->';
```
Why do we use `<details>` here? It isn't used by WordPress Accordion block as well as the template returned by `getTemplate`.

dinhtungdu reviewed (changes requested):
Thanks for the PR! I can confirm it is working as expected for default cases. However, I encountered an issue while trying to extend the block using Block Hook. Adding more context to my comment below.

When I tried the snippet in https://github.com/woocommerce/woocommerce/pull/57893, here is what I got, the hooked block isn't rendered correctly. Can you take a look at this?

<img width="2885" height="1679" alt="Image" src="https://github.com/user-attachments/assets/4cc361cf-da43-4495-831a-ecfc88e9a543" />

AbdullahAymanMSRE committed:
Refactor Product Details block to use div elements for accordion items and update test cases accordingly

- Change accordion item structure from <details> to <div> for better accessibility and styling.
- Update related test cases to reflect the new structure and ensure correct visibility of additional information.
- Fix PHP linting issues and improve code formatting.
- Fix client e2e tests.

AbdullahAymanMSRE committed:
Update e2e tests for Add to Cart + Options Block to handle WordPress version differences

- Ensure backward compatibility with earlier versions by adjusting the selector logic.

dinhtungdu commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductDetails.php:
```diff
@@ -162,22 +163,43 @@ function ( $a, $b ) {
-			<!-- /wp:woocommerce/accordion-panel --></div>
-			<!-- /wp:woocommerce/accordion-item -->';
+		if ( Utils::wp_version_compare( '6.9', '>=' ) ) {
+			$template = '<!-- wp:core/accordion-item -->
```
Thanks for the update, change tests good, and I can see the hooked block as expected. However, there is a small issue that I may have missed during previous reviews here: we shouldn't use `core/` prefix for Core blocks. It's not wrong, but it's the convention for core blocks. Can you please remove it? Thanks!

AbdullahAymanMSRE committed:
Remove core/ prefix from accordion block markup to follow WordPress conventions.

dinhtungdu reviewed (commented):
@AbdullahAymanMSRE, can you please look at the failed E2E test and fix it? The failed test was updated in this PR, so it's not a flaky one.

AbdullahAymanMSRE committed:
Merge remote-tracking branch 'upstream/trunk' into fix/product-details-use-core-accordion

AbdullahAymanMSRE committed:
Add wpCoreVersion to e2e tests for Add to Cart + Options Block

dinhtungdu committed:
fix: using anchor block instead of WP version to inject hooked block

dinhtungdu commented:
@AbdullahAymanMSRE Thanks for the update. In my final review of this PR, I found it incorrect to determine the version of the accordion block using the WordPress version. Instead, we should use the anchor block to determine which accordion block we should use. Because there may be stores upgrading from 6.8 to 6.9 and still using the WooCommerce version of the accordion block. I updated that behavior in 76079a7. @Aljullu @AbdullahAymanMSRE can you take a look at it?

dinhtungdu committed:
chore: phpcs

dinhtungdu commented:
@AbdullahAymanMSRE  It's all good now. I merged your PR into trunk. Thanks again for the contribution!
